### PR TITLE
Add Pad1d functions and ConstantPad3d

### PR DESCRIPTION
**Pull request by dhpollack**
Related to #1825 and #2637 

I've added the 1d padding functions that were missing and the 3d constant value padding.  It's not quite a generalized solution as suggested in 1825, but it does plug some of the holes in the current functionality of the padding functions.

### CODE REVIEW

**Comment by soumith** [comment]
we'll look into the contbuild failure, looks unrelated to your PR.

**Comment by dhpollack** [comment]
the second commit does Nd constant padding.  I do have a question.  I am allowing for padding on all dimensions when using the torch.nn.functional.pad function as long as the length of the padding tuple / 2 is less than the length of the tensor dimension.  But should I be excluding the first two dimensions?  For the ConstantPad1/2/3d functions the first two dimensions are excluded.  Only with the functional would one be allowed to pad in any dimension.

File: torch/nn/_functions/padding.py [hunk 1]
```diff
@@ -1,196 +1,45 @@
...
-
-        return grad_input, None, None
-
-class ConstantPad2d(Function):
+class ConstantPadNd(Function):
 
     @staticmethod
     def forward(ctx, input, pad, value=0):
-        assert input.dim() == 4, 'only 4D supported for padding'
+        assert len(pad) % 2 == 0, "padding tuple must be an even length"
```
Comment (soumith) [inline torch/nn/_functions/padding.py]: dont you additionally have to assert that `len(pad) / 2 == (input.dim() - 2)`

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1065,36 +1065,34 @@ def pad(input, pad, mode='constant', value=0):
...
     Args:
         input (Variable): 3D, 4D or 5D tensor
-        pad (tuple): 2-, 4-, or 6-elem tuple
+        pad (tuple): m-elem tuple, where m // 2 > input dimension and m % 2 == 0
         mode: 'constant', 'reflect' or 'replicate'. Default: 'constant'
         value: fill value for 'constant' padding. Default: 0
     """
-    if input.dim() == 3:
+    assert len(pad) % 2 == 0, 'padding length must be divisible by 2'
+    assert len(pad) // 2 <= len(input.size()), 'padding length too large'
```
Comment (soumith) [inline torch/nn/functional.py]: oh i see, you've asserted here, so it's not really necessary in the `Function`

File: torch/nn/modules/padding.py [hunk 3]
```diff
@@ -28,7 +60,7 @@ class ConstantPad1d(Module):
     """
 
     def __init__(self, padding, value):
-        super(ConstantPad1d, self).__init__()
+        super(ConstantPad1d, self).__init__(padding, value)
```
Comment (soumith) [inline torch/nn/modules/padding.py]: What's the value of subclassing these classes on `(_ConstantPadNd)` if you are not actually doing any code de-duplication.

Comment (soumith) [reply torch/nn/modules/padding.py]: Also, we kept around these nn classes only for legacy purposes, after the F.pad refactor. Going forward, `F.pad` is the blessed and nice way to go forward.

**Comment by soumith** [comment]
> I am allowing for padding on all dimensions when using the torch.nn.functional.pad function as long as the length of the padding tuple / 2 is less than the length of the tensor dimension. But should I be excluding the first two dimensions?

Preserve backward compatibility. If F.pad excluded the first two dims, keep it as such. If F.pad includes first two dims, keep it as such.

The PR looks largely bug-free and good to go.

**Comment by soumith** [comment]
Oh, i see what you've done: if you give 4D Tensor and 2D padding tuple, that's fine. if you give 4D Tensor (NCHW) and 3D padding tuple, then it pads CHW, and if you give 4D Tensor and 4D padding tuple, then it pads all 4 dims.

This seems reasonable, as long as it's well documented. The documentation for F.pad has to be improved do doc this, and we should be good to go.

File: torch/nn/modules/padding.py (hunk 3 above)
Comment (dhpollack) [reply torch/nn/modules/padding.py]: I originally used the torch/nn/modules/conv.py as a template, but you are right, I could just make each of these a module and get rid of _ConstantPadNd.

**Comment by dhpollack** [comment]
Yea, I believe this is "backwards compatible".  Unless you had code that didn't run before, i.e. a 4D tensor trying to pad one dimension.  So if you had a tensor of size = (2, 3, 4, 5) and of padding=(1, 1) it would fail before, but now it would return a tensor of size (2, 3, 4, 7).  But the behavior of F.pad(size(2, 3, 4, 5), (1, 1, 1, 1)) will return the same tensor as before. 

Ok, I'll work on the documentation. It is a little confusing to explain because the padding tuple goes in reverse order of the dimensions.  In the nn/modules/padding.py file, I used the following explanation.  Would you prefer something like this or more like what you've done with using the 4D tensor as an example and then letting the user intuitively figuring out what happens in N-dimensions.  I sided on being more mathy here, but actually find it sort of difficult to understand myself.

    r"""Pads the input tensor boundaries with a constant value.

    Args:
        padding (int, tuple): the size of the padding. If is int, uses the same
            padding in both boundaries. If a m-tuple, where m is divisible by
            2, then padding is(begin_dim_pad, end_dim_pad, ...) for each pair
            paddings starting from the last dimension of the input moving
            forward.
        value (int): value inserted into tensor as padding

    Shape:
        - Input: :math:`(N_i, D_j_{in})`
        - Output: :math:`(N_i, D_j_{out})` where
          :math:`J = len(padding) // 2`
          :math:`j = (0, 1, ..., J)`
          :math:`~j = (J, J-1, ..., 0)`
          :math:`D_j_{out} = D_j_{in} + padding[~j*2] + padding[~j*2 + 1]`

    """

File: torch/nn/modules/padding.py (hunk 3 above)
Comment (dhpollack) [reply torch/nn/modules/padding.py]: Ok, so if nn.ConstantPad1d and nn.ConstantPad2d did not exist, should I remove those?  Also, should there be a message in the docs that the functional should be used instead of these modules so people know where to look if they want to use "nn.ConstantPad1/3d"?

Comment (soumith) [reply torch/nn/modules/padding.py]: yes, remove them if they didn't exist, and for the other padding layers, add a not in the docs saying F.pad is preferred.

**Comment by soumith** [comment]
yea the current blurb is a bit confusing.
> If a m-tuple, where m is divisible by
        2, then padding is(begin_dim_pad, end_dim_pad, ...) for each pair
        paddings starting from the last dimension of the input moving
        forward.

However, whenever something like this is written, showing two or three examples gives a lot more clarity. So showcase the case where 3D input Tensor with 1D tuple, 2D tuple and 3D tuple.
I think that'll be sufficient.

**Comment by soumith** [comment]
Thanks a lot @dhpollack . Your PRs here and in pytorch/audio are quite awesome!!!

**Comment by GilbertHeywood** [comment]
For consistency, could you also add ZeroPad1d? Though I know ConstantPad1d covers that.
